wgengine/netstack/netstack.go

```diff
@@ -381,18 +381,53 @@ func (ns *Impl) acceptUDP(r *udp.ForwarderRequest) {
 		ns.logf("Could not create endpoint, exiting")
 		return
 	}
+	localAddr, err := ep.GetLocalAddress()
+	if err != nil {
+		return
+	}
 	c := gonet.NewUDPConn(ns.ipstack, &wq, ep)
-	go echoUDP(c)
+	go ns.forwardUDP(c, &wq, localAddr)
 }
 
-func echoUDP(c *gonet.UDPConn) {
-	buf := make([]byte, 1500)
-	for {
-		n, err := c.Read(buf)
-		if err != nil {
-			break
+func (ns *Impl) forwardUDP(client *gonet.UDPConn, wq *waiter.Queue, localAddr tcpip.FullAddress) {
```

Comment:
You're basically copy/pasting the TCP code here, but the model is different.

I don't think you want to make a whole "session" out of something that doesn't really have a session.

A UDP connection never has a "close" event to wait on, so your io.Copy calls will run forever. Also, io.Copy on UDP doesn't necessarily preserve packet boundaries which are important in UDP. It's not a byte stream.

Can we instead just send individual packets as they arrive? Maybe using IP raw connections?

https://pkg.go.dev/golang.org/x/net/ipv4/#RawConn perhaps?